File: core-scala/src/main/java/io/asterisque/core/session/Dispatcher.java
package io.asterisque.core.session;

import io.asterisque.Asterisque;
import io.asterisque.core.Debug;
import io.asterisque.core.codec.VariableCodec;
import io.asterisque.core.msg.Control;
import io.asterisque.core.msg.Message;
import io.asterisque.core.msg.SyncSession;
import io.asterisque.core.wire.MessageQueue;
import io.asterisque.core.wire.Wire;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.net.ssl.SSLException;
import javax.net.ssl.SSLSession;
import java.util.Arrays;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicReference;

public class Dispatcher {
  private static final Logger logger = LoggerFactory.getLogger(Dispatcher.class);

  /**
   * このディスパッチャーのノード ID
   */
  private final UUID nodeId;

  private final ConcurrentHashMap<UUID, Session> sessions = new ConcurrentHashMap<>();

  private final Executor executor;

  private final VariableCodec codec;

  public Dispatcher(@Nonnull UUID nodeId, @Nonnull Executor executor, @Nonnull VariableCodec codec) {
    this.nodeId = nodeId;
    this.executor = executor;
    this.codec = codec;
  }

  /**
   * このディスパッチャー上で提供するサービス。
   */
  public final Services services = new Services() {
    private final ConcurrentHashMap<String, Service> services = new ConcurrentHashMap<>();

    @Override
    @Nullable
    public Service set(@Nonnull String serviceId, @Nonnull Service service) {
      return services.put(serviceId, service);
    }

    @Override
    @Nullable
    public Service remove(@Nonnull String serviceId) {
      return services.remove(serviceId);
    }

    @Override
    @Nullable
    public Service get(@Nonnull String serviceId) {
      return services.get(serviceId);
    }

  };

  public CompletableFuture<Session> bind(@Nonnull Wire wire, @Nonnull String serviceId, int ping, int sessionTimeout) {

    // セッション ID の決定
    UUID sessionId = Asterisque.Zero;
    if (wire.isPrimary()) {
      do {
        sessionId = UUID.randomUUID();
      } while (sessions.containsKey(sessionId));
      logger.trace("new session-id issued: {}", sessionId);
    }

    // ハンドシェイクの開始
    Handshake handshake = new Handshake(wire, sessionId, serviceId, ping, sessionTimeout);
    return handshake.future;
  }

  @Nonnull
  private Session handshake(@Nonnull Wire wire, @Nonnull SyncSession sent, @Nonnull SyncSession received) {
    String logPrifix = Asterisque.logPrefix(wire.isPrimary(), sent.sessionId);

    // TODO クライアント証明が必要な場合はこの位置で相手の SYNC_SESSION と wire の証明書を検証
    wire.session().ifPresent(tls -> {
      logger.debug("{}: verifying ssl certificate: {}", logPrifix, tls);
      if(logger.isTraceEnabled()){
        try {
          Debug.dumpSSLSession(logger, logPrifix, tls);
        } catch(SSLException ex){
          logger.warn("{}: cannot verify ssl certificate: {}", logPrifix, tls);
        }
      }
      // TODO 証明書の検証が正しいことを確認
      // TODO 証明書の検証を行うかや、ローカルインストール済みの CA 証明書なども検証
      if(! tls.isValid() || !Arrays.equals(tls.getId(), received.nodeId.toString().getBytes())){
        throw new ProtocolViolationException("invalid ssl certificate");
      }
    });

    SyncSession.Pair sync = wire.isPrimary() ? new SyncSession.Pair(sent, received) : new SyncSession.Pair(received, sent);
    UUID sessionId = sync.sessionId();
    AtomicReference<Session> ref = new AtomicReference<>();
    sessions.compute(sessionId, (ignored, oldSession) -> {
      if (sessionId.equals(Asterisque.Zero) || oldSession != null) {
        logger.debug("{}: session id collision: {}", logPrifix, sessionId);
        wire.outbound.offer(new Control(Control.Close));
        wire.close();
        return oldSession;
      }
      Session session = new Session(sessionId, this, wire, codec, sync);
      session.addListener(session1 -> {
        logger.debug("{}: session closed: {}", logPrifix, sessionId);
        sessions.remove(sessionId);
      });
      ref.set(session);
      return ref.get();
    });
    logger.info("{}: handshake success, beginning session", logPrifix);
    return ref.get();
  }

  CompletableFuture<Object> dispatch(String serviceId, @Nonnull Pipe pipe, @Nonnull String logId) {
    Service service = services.get(serviceId);
    if (service == null) {
      CompletableFuture<Object> future = new CompletableFuture<>();
      String msg = String.format("no such service: %s", serviceId);
      logger.debug("{}: {}", logId, msg);
      future.completeExceptionally(new NoSuchServiceException(msg));
      return future;
    }
    return service.apply(pipe, executor);
  }

  public interface Services {
    @Nullable
    Service set(@Nonnull String serviceId, @Nonnull Service service);

    @Nullable
    Service remove(@Nonnull String services);

    @Nullable
    Service get(@Nonnull String serviceId);
  }

  private class Handshake implements MessageQueue.Listener {
    public final CompletableFuture<Session> future;
    private final Wire wire;
    private final SyncSession sent;
    private boolean polled = false;

    private Handshake(Wire wire, UUID sessionId, String serviceId, int ping, int sessionTimeout) {
      this.future = new CompletableFuture<>();
      this.wire = wire;

      // ハンドシェイクメッセージの送信
      this.sent = new SyncSession(nodeId, sessionId, serviceId, System.currentTimeMillis(), ping, sessionTimeout);
      wire.outbound.offer(sent.toControl());

      // ハンドシェイクメッセージの受信ハンドラを設置
      wire.inbound.addListener(this);
    }

    @Override
    public void messagePollable(@Nonnull MessageQueue messageQueue, boolean pollable) {

      // 受信メッセージの取得
      Message msg = pollable && !polled ? messageQueue.poll() : null;
      if (msg == null) {
        return;
      }
      polled = true;

      try {
        if (msg instanceof Control) {
          future.complete(handshake(wire, sent, SyncSession.parse((Control) msg)));
        } else {
          String message = String.format("SyncSession control message expected but a different one detected: %s", msg);
          logger.error(message);
          throw new ProtocolViolationException(message);
        }
      } catch (Exception ex) {
        logger.error(ex.toString());
        future.completeExceptionally(ex);
        wire.close();
      } finally {
        messageQueue.removeListener(this);
        // TODO ハンドシェイクタイムアウト設定
      }
    }
  }

}

File: core-scala/src/main/scala/io/asterisque/wire/rpc/Dispatcher.scala
package io.asterisque.wire.rpc

import java.security.cert.X509Certificate
import java.util.Objects
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

import io.asterisque.wire.gateway.{MessageQueue, Wire}
import io.asterisque.wire.message.Message.Control
import io.asterisque.wire.message.SyncSession
import io.asterisque.wire.rpc.Dispatcher._
import io.asterisque.wire.{AuthenticationException, ProtocolException}
import javax.annotation.{Nonnull, Nullable}
import org.slf4j.LoggerFactory

import scala.annotation.tailrec
import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.util.Random

/**
  * 複数のサービスを保持し、リモートからのセッション開始要求に応える
  *
  * @param context ディスパッチャーがリクエストを処理するためのスレッドプール
  */
class Dispatcher private[rpc](val context:ExecutionContext, codec:Codec) {

  /**
    * このディスパッチャー上でサービスの処理を行っているセッション。
    */
  private[this] val sessions = new ConcurrentHashMap[Long, Session]()

  /**
    * 新規セッションを識別するための ID を取得するためのシーケンス。
    */
  private[this] val sessionIdSequence = new AtomicLong(0)

  /**
    * このディスパッチャー上で提供するサービス。
    */
  object services {
    private[this] val services = new ConcurrentHashMap[String, Service]()

    @Nullable
    def set(@Nonnull serviceId:String, @Nonnull service:Service):Service = {
      Objects.requireNonNull(serviceId)
      Objects.requireNonNull(service)
      services.put(serviceId, service)
    }

    @Nullable
    def remove(@Nonnull serviceId:String):Service = {
      Objects.requireNonNull(serviceId)
      services.remove(serviceId)
    }

    @Nullable
    def get(@Nonnull serviceId:String):Service = {
      Objects.requireNonNull(serviceId)
      services.get(serviceId)
    }
  }

  def bind(@Nonnull wire:Wire, conf:Map[String, String]):Future[Session] = {

    // TLS 証明書でのピア認証が必要な場合は TLS Session の中から信頼済み証明書を参照
    val validCertificates:Seq[X509Certificate] = wire.session.map { tls =>
      // NOTE TLS で TrustManager による認証が通過しているためこの証明書パスは信頼可能
      tls.getPeerCertificates.map(_.asInstanceOf[X509Certificate]).toSeq
    }.getOrElse {
      val msg = s"peer authentication is required but not by TLS"
      return Future.failed(new AuthenticationException(msg))
    }

    // セッション ID の決定
    var sessionId:Long = 0
    if(wire.isPrimary) {
      do {
        sessionId = Random.nextLong()
      } while(sessions.containsKey(sessionId))
      logger.trace("new session-id issued: {}", sessionId)
    }

    // ハンドシェイクの開始
    new Handshake(wire, conf, validCertificates).future
  }

  @Nonnull
  private def handshake(@Nonnull hs:Handshake, @Nonnull received:SyncSession):Future[Session] = {

    // 新規セッションの構築と登録
    @tailrec
    def _newSession():Session = {
      val sessionId = sessionIdSequence.getAndIncrement()
      val oldSession = sessions.computeIfAbsent(sessionId, { _:Long =>
        val pair = if(hs.wire.isPrimary) SyncSession.Pair(hs.sent, received) else SyncSession.Pair(received, hs.sent)
        val session = new Session(sessionId, this, hs.wire, codec, pair)
        session.addListener { _:Session =>
          logger.debug("{}: session closed: {}", session.logId, sessionId)
          sessions.remove(sessionId)
        }
        session
      })
      if(oldSession == null) sessions.get(sessionId) else _newSession()
    }

    val newSession = _newSession()
    logger.info("{}: handshake success, beginning session", newSession.logId)
    Future.successful(newSession)
  }

  /**
    * 指定されたサービスの処理を呼び出します。
    *
    * @param pipe  呼び出しのパイプ
    * @param logId ログ出力用の文字列
    * @return サービスの呼び出し結果
    */
  private[rpc] def dispatch(@Nonnull pipe:Pipe, @Nonnull logId:String):Future[Array[Byte]] = {
    val service = services.get(pipe.service)
    if(service == null) {
      val msg = s"no such service: ${pipe.service}"
      logger.debug(s"$logId: $msg")
      Future.failed(new ProtocolException(msg))
    } else {
      service.apply(codec, pipe, context)
    }
  }

  /**
    * メッセージキューに対して [[SyncSession]] ハンドシェイクの到着を待機するクラスです。
    *
    * @param wire SyncSession の到着を待機する Wire
    * @param conf セッション同期の設定
    */
  private[this] class Handshake(val wire:Wire, val conf:Map[String, String], val validCerts:Seq[X509Certificate]) extends MessageQueue.Listener {

    private[this] val promise = Promise[Session]()

    val sent:SyncSession = SyncSession(System.currentTimeMillis, conf)

    def future:Future[Session] = promise.future

    // ハンドシェイクメッセージの受信ハンドラを設置
    wire.inbound.addListener(this)

    // ハンドシェイクメッセージの送信
    wire.outbound.offer(Control(sent))

    override def messagePollable(@Nonnull messageQueue:MessageQueue, pollable:Boolean):Unit = if(pollable) {
      // 受信メッセージの取得
      val msg = messageQueue.poll()
      if(msg != null && !promise.isCompleted) {
        messageQueue.removeListener(this)
        msg match {
          case Control(sync:SyncSession) =>
            promise.completeWith(handshake(this, sync))
          case _ =>
            val message = s"SyncSession control message expected but a different one detected: $msg"
            logger.error(message)
            promise.failure(new ProtocolException(message))
            wire.close()
        }
      }
    }
  }

}

object Dispatcher {
  private val logger = LoggerFactory.getLogger(classOf[Dispatcher])

  trait Services {
    @Nullable
    def set(@Nonnull serviceId:String, @Nonnull service:Service):Service

    @Nullable
    def remove(@Nonnull services:String):Service

    @Nullable
    def get(@Nonnull serviceId:String):Service
  }

}
